### PR TITLE
fix: restart Cypress server and browser on baseUrl change

### LINKED ISSUE #22056: Changing the baseUrl value in cypress.config.js will not take effect until restarting cypress
### Current behavior

when changing the value of `baseUrl` in `cypress.config.js` both the cypress window as well as the cypress controlled instance of the browser reload giving the impression that the changes to configuration are taking effect, this is not the case however

Note: I did not test all values that can be altered in `cypress.config.js` however at least with `viewportWidth` and `viewportHeight`, altering them and saving does not require a restart of cypress for them to take effect. 

### Desired behavior

_No response_

### Test code to reproduce

create cypress.config.js with the following code:
```js
const { defineConfig } = require('cypress')

module.exports = defineConfig({

  e2e: {
    baseUrl: 'https://www.google.com/'
  }
})
```
`cypress open` in terminal 
change the contents of cypress.config.js to the following while cypress is running:

```js
const { defineConfig } = require('cypress')

module.exports = defineConfig({

  e2e: {
    baseUrl: 'https://www.cypress.io/'
  }
})
```

you will notice that even though the page reloads the test still go to the first url

### Cypress Version

v10.0.1

### Other

_No response_

### PR DESCRIPTION
**Pull request by ryanthemanuel**
- Closes #22056 

### User facing changelog

The `baseUrl` config option will now restart the Cypress server automatically when it is updated in a Cypress config file.

### Additional details

There are several things that use the baseUrl during the server's initialization like the [browser and xhr urls](https://github.com/cypress-io/cypress/blob/develop/packages/server/lib/config.ts#L532-L534) and [remote states](https://github.com/cypress-io/cypress/blob/develop/packages/server/lib/server-e2e.ts#L128). In order to properly ensure they are reinitialized appropriately, we will need to restart the server. Thus, we change `baseUrl`'s `requireRestartOnChange` property to `server`.

### Steps to test

1. `cypress open` a project with a `baseUrl`
2. Execute a test within the project
3. Update the `baseUrl` in cypress.config.js
4. Ensure that the server/browser relaunches and that the url in the browser references the updated `baseUrl`

### How has the user experience changed?

Before

https://user-images.githubusercontent.com/4873279/172271308-5561a2e3-d6a1-41af-8e24-721dc6e70c01.mov

After

https://user-images.githubusercontent.com/4873279/172271323-bcdf97b9-9ed0-4cb6-b749-f030306dc5f1.mov

### PR Tasks

- [X] Have tests been added/updated?
- [X] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/launchpad/cypress/e2e/choose-a-browser.cy.ts [hunk 1]
```diff
@@ -190,7 +190,7 @@ describe('Choose a Browser Page', () => {
       cy.get('h1').should('contain', 'Choose a Browser')
```
Comment (ryanthemanuel) [inline packages/launchpad/cypress/e2e/choose-a-browser.cy.ts:190]: These changes are for unrelated flake.

File: packages/launchpad/cypress/e2e/choose-a-browser.cy.ts [hunk 2]
```diff
@@ -220,7 +220,7 @@ describe('Choose a Browser Page', () => {
 
       cy.visitLaunchpad()
 
-      cy.withCtx((ctx) => {
+      cy.withCtx((ctx, o) => {
```
Comment (ryanthemanuel) [inline packages/launchpad/cypress/e2e/choose-a-browser.cy.ts:223]:
```suggestion
      cy.withCtx((ctx) => {
```

File: packages/launchpad/cypress/e2e/choose-a-browser.cy.ts [hunk 3]
```diff
@@ -276,9 +276,13 @@ describe('Choose a Browser Page', () => {
...
       cy.findByRole('radio', { name: 'Chrome v1', checked: true }).as('chromeItem')
 
+      cy.withCtx((ctx, o) => {
+        o.sinon.stub(ctx.actions.project, 'launchProject')
+      })
+
       cy.contains('button', 'Start E2E Testing in Chrome').should('be.visible').click()
 
-      cy.withCtx((ctx) => {
+      cy.withCtx((ctx, o) => {
```
Comment (ryanthemanuel) [inline packages/launchpad/cypress/e2e/choose-a-browser.cy.ts:285]:
```suggestion
      cy.withCtx((ctx) => {
```

**Review by tgriesser — APPROVED** [review]
Nice - changes look pretty straightforward to me

**Review by tbiethman — APPROVED** [review]
Works as advertised :+1:

**Comment by ZachJW34** [comment]
Just wanted to call out https://github.com/cypress-io/cypress/issues/22072. We have [baseUrl](https://docs.cypress.io/guides/references/configuration#Environment-Variables) as a property that can be changed at the test level, but with these changes it looks like the documentation would be incorrect.

**Review by ZachJW34 — APPROVED** [review]

**Review by lmiller1990 — APPROVED** [review]

**Pull request merged**